beacon-chain/core/blocks/withdrawals.go

```diff
@@ -148,31 +149,60 @@ func ValidateBLSToExecutionChange(st state.ReadOnlyBeaconState, signed *ethpb.Si
 //	        next_validator_index = ValidatorIndex(next_index % len(state.validators))
 //	        state.next_withdrawal_validator_index = next_validator_index
 func ProcessWithdrawals(st state.BeaconState, executionData interfaces.ExecutionData) (state.BeaconState, error) {
+	if st.Version() >= version.EPBS {
+		IsParentBlockFull, err := st.IsParentBlockFull()
+		if err != nil {
+			return nil, errors.Wrap(err, "could not check if parent block is full")
+		}
+
+		if !IsParentBlockFull {
+			return nil, nil
+		}
+	}
+
 	expectedWithdrawals, partialWithdrawalsCount, err := st.ExpectedWithdrawals()
 	if err != nil {
 		return nil, errors.Wrap(err, "could not get expected withdrawals")
 	}
 
 	var wdRoot [32]byte
-	if executionData.IsBlinded() {
-		r, err := executionData.WithdrawalsRoot()
+	if st.Version() >= version.EPBS {
+		withdrawal_list := make([]*enginev1.Withdrawal, 0, fieldparams.MaxWithdrawalsPerPayload)
+		withdrawal_list = append(withdrawal_list, expectedWithdrawals...)
+		wdRoot, err = ssz.WithdrawalSliceRoot(withdrawal_list, fieldparams.MaxWithdrawalsPerPayload)
+		if err != nil {
+			return nil, errors.Wrap(err, "could not hash withdrawals list")
+		}
+		err = st.SetLastWithdrawalsRoot(wdRoot[:])
+		if err != nil {
+			return nil, errors.Wrap(err, "could not set withdrawals root")
+		}
+		r, err := st.LastWithdrawalsRoot()
 		if err != nil {
 			return nil, errors.Wrap(err, "could not get withdrawals root")
 		}
 		wdRoot = bytesutil.ToBytes32(r)
 	} else {
-		wds, err := executionData.Withdrawals()
-		if err != nil {
-			return nil, errors.Wrap(err, "could not get withdrawals")
-		}
+		if executionData.IsBlinded() {
+			r, err := executionData.WithdrawalsRoot()
+			if err != nil {
+				return nil, errors.Wrap(err, "could not get withdrawals root")
+			}
+			wdRoot = bytesutil.ToBytes32(r)
+		} else {
+			wds, err := executionData.Withdrawals()
+			if err != nil {
+				return nil, errors.Wrap(err, "could not get withdrawals")
+			}
 
-		if len(wds) != len(expectedWithdrawals) {
-			return nil, fmt.Errorf("execution payload header has %d withdrawals when %d were expected", len(wds), len(expectedWithdrawals))
-		}
+			if len(wds) != len(expectedWithdrawals) {
+				return nil, fmt.Errorf("execution payload header has %d withdrawals when %d were expected", len(wds), len(expectedWithdrawals))
+			}
 
-		wdRoot, err = ssz.WithdrawalSliceRoot(wds, fieldparams.MaxWithdrawalsPerPayload)
-		if err != nil {
-			return nil, errors.Wrap(err, "could not get withdrawals root")
+			wdRoot, err = ssz.WithdrawalSliceRoot(wds, fieldparams.MaxWithdrawalsPerPayload)
```

Comment:
In line 209 you are hashing again the same list of withdrawals that you hashed above in the ePBS path. You can move up that hash and avoid repeating it.